src/AggregateFunctions/AggregateFunctionGroupArrayIntersect.cpp

```diff
@@ -0,0 +1,466 @@
+
+    AggregateFunctionGroupArrayIntersect(const DataTypePtr & argument_type, const Array & parameters_, const DataTypePtr & result_type_)
+        : IAggregateFunctionDataHelper<AggregateFunctionGroupArrayIntersectData<T>,
+          AggregateFunctionGroupArrayIntersect<T>>({argument_type}, parameters_, result_type_) {}
+
+    String getName() const override { return "GroupArrayIntersect"; }
+
+    bool allocatesMemoryInArena() const override { return false; }
+
+    void add(AggregateDataPtr __restrict place, const IColumn ** columns, size_t row_num, Arena *) const override
+    {
+        auto & version = this->data(place).version;
+        auto & map = this->data(place).value;
+
+        const auto data_column = assert_cast<const ColumnArray &>(*columns[0]).getDataPtr();
+        const auto & offsets = assert_cast<const ColumnArray &>(*columns[0]).getOffsets();
+        const size_t offset = offsets[static_cast<ssize_t>(row_num) - 1];
+        const auto arr_size = offsets[row_num] - offset;
+
+        ++version;
+        if (version == 1)
+        {
+            for (size_t i = 0; i < arr_size; ++i)
+                map[static_cast<T>((*data_column)[offset + i].get<T>())] = version;
+        }
+        else if (map.size() > 0)
+        {
+            typename State::Map new_map;
+            for (size_t i = 0; i < arr_size; ++i)
+            {
+                typename State::Map::LookupResult value = map.find(static_cast<T>((*data_column)[offset + i].get<T>()));
+                if (value != nullptr && value->getMapped() == version - 1)
+                    new_map[static_cast<T>((*data_column)[offset + i].get<T>())] = version;
+            }
+            map = std::move(new_map);
+        }
+    }
+
+    void merge(AggregateDataPtr __restrict place, ConstAggregateDataPtr rhs, Arena *) const override
+    {
+        auto & map = this->data(place).value;
+        const auto & rhs_map = this->data(rhs).value;
+
+        if (this->data(rhs).version == 0)
+            return;
+
+        UInt64 version = this->data(place).version++;
+        if (version == 0)
+        {
+            for (auto & rhs_elem : rhs_map)
+                map[rhs_elem.getKey()] = 1;
+            return;
+        }
+
+        if (map.size() > 0)
+        {
+            typename State::Map new_map;
+            if (rhs_map.size() < map.size())
+            {
+                for (auto & rhs_elem : rhs_map)
+                {
+                    auto value = map.find(rhs_elem.getKey());
+                    if (value != nullptr)
+                        new_map[rhs_elem.getKey()] = version;
+                }
+            }
+            else
+            {
+                for (auto & elem : map)
+                {
+                    auto value = rhs_map.find(elem.getKey());
+                    if (value != nullptr)
+                        new_map[elem.getKey()] = version;
+                }
+            }
```

Comment:
Since the branches are symmetrical:
``` cpp
auto create_new_map = [](auto & lhs, auto & rhs) { ... };
auto new_map = rhs_map_.size() < map.size() ? create_new_map(rhs_map, map) : create_new_map(map, rhs_map);
```